### ISSUE TITLE
False positive for JSX props with conditional types

### ISSUE THREAD

**Issue by marcospassos** [microsoft/TypeScript#49476]
# Bug Report
Typescript can't recognize correctly recognize conditional types passed through JSX props.

### 🔎 Search Terms

- Spread generics
- Spread react generics

### 🕗 Version & Regression Information

4.72

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.4#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDez0wCiANkiEgHZZwBiUKA5uVQL5wBmUEIcA5FKgy8A3AChRMAJ5gktCBAAKXMAGc4AXlyi4OjvIBccFTCjAKDTKJZiJ02YQAeJlDXlKIqgDwK4SJ5QATNVdFZRUAPg04AHkQYBhvbABrJEkIdjlQjwibKRk4AGEAV2NuABFgFTBiFEl3Lx8-GEDgtzDIzQUUWGAUYk9HZxD6lW9wyKaWuEH6YbCx7V0AflwwMKXDGZc27LGWRZ1DHDXszadZnYbw63F2IooMYAgKOAAhTwAVX38KIMyR8IAChOqkMxVKIAqVRqdXmH3CAEpDPgMCQyJR4DgDnABDAilAXp46IxmPAAPThMT7UR3B4wJ4vACCn2+zV+rSyqiBIJUYJKMHKlWqtRGn0RyMERFIpK0uhxSDxBLgnleuAAdBqeWwKVSgA)

### 💻 Code

```ts
import React, {ReactElement, Fragment} from 'react';

type FooProps = {
    foo: string,
};

type ExtraFooProps<P extends FooProps> = Omit<P, keyof FooProps>;

type CustomDisplayProps<P extends FooProps> = Partial<ExtraFooProps<P>> extends ExtraFooProps<P>
    ? {props?: ExtraFooProps<P>}
    : {props: ExtraFooProps<P>};

function B<T extends FooProps>(props: CustomDisplayProps<T>): ReactElement {
    return <Fragment />;
}

function A<T extends FooProps>(props: CustomDisplayProps<T>): ReactElement {
    return <B {...props} />;
}
```

### 🙁 Actual behavior

It doesn't recognize the props are valid, even though the props and signature are the same.

### 🙂 Expected behavior

No errors.

**Comment by RyanCavanaugh** [comment]
The condition TS is trying to warn you about is somewhat valid. The issue here is that a potential instantiation of `A` includes a type with all-optional members, none of which match any *visible* property of the target type. The problem is that the set of *visible properties* in this case is just the intrinsic JSX attributes (key, ref, etc) and none of the `FooProps` due to `B`'s props being conditional on a generic type. That said, if these generics *did anything*, it'd be possible to construct a more compelling example.

Workaround:
```ts
return <B<FooProps> {...props} />;
```

**Comment by marcospassos** [comment]
@RyanCavanaugh, I still get the error even if not all types are part of the conditional. The example is just a simplification of the real type. The proposed workaround does not work for such cases:

```ts
import React, {ReactElement, Fragment} from 'react';

type FooProps = {
    foo: string,
};

type ExtraFooProps<P extends FooProps> = Omit<P, keyof FooProps>;

type CustomDisplayProps<P extends FooProps> = {
    example: string,
    } & (
        Partial<ExtraFooProps<P>> extends ExtraFooProps<P>
        ? {props?: ExtraFooProps<P>}
        : {props: ExtraFooProps<P>}
    );

function B<T extends FooProps>(props: CustomDisplayProps<T>): ReactElement {
    return <Fragment />;
}

function A<T extends FooProps>(props: CustomDisplayProps<T>): ReactElement {
    return <B<T> {...props} />;
}
```

**Comment by marcospassos** [comment]
The best alternative so far:

```tsx
function A<T extends FooProps>(props: CustomDisplayProps<T>): ReactElement {
    return <B  {...(props as typeof props & JSX.IntrinsicAttributes)}/>;
}
```

Validation works, at least.
